zio-json/shared/src/main/scala/zio/json/uuid/UUIDParser.scala

```diff
@@ -0,0 +1,154 @@
+        val lsb4 = parseNibbles(input, 32)
+        if ((msb1 | msb2 | msb3 | msb4 | lsb1 | lsb2 | lsb3 | lsb4) >= 0) {
+          new java.util.UUID(
+            msb1 << 48 | msb2 << 32 | msb3 << 16 | msb4,
+            lsb1 << 48 | lsb2 << 32 | lsb3 << 16 | lsb4
+          )
+        } else unsafeParseExtended(input)
+      } else unsafeParseExtended(input)
+    } else unsafeParseExtended(input)
+
+  // A nibble is 4 bits
+  @nowarn("msg=implicit numeric widening")
+  private def parseNibbles(input: String, position: Int): Long = {
+    val ns  = HexToDecimal
+    val ch1 = input.charAt(position)
+    val ch2 = input.charAt(position + 1)
+    val ch3 = input.charAt(position + 2)
+    val ch4 = input.charAt(position + 3)
+
+    if ((ch1 | ch2 | ch3 | ch4) > 0xff) -1L
+    else ns(ch1) << 12 | ns(ch2) << 8 | ns(ch3) << 4 | ns(ch4)
+  }
+
+  private def unsafeParseExtended(name: String): java.util.UUID = {
+    val len = name.length
+    if (len > 36) throw new IllegalArgumentException("UUID string too large")
+    val dash1 = name.indexOf('-', 0)
+    val dash2 = name.indexOf('-', dash1 + 1)
+    val dash3 = name.indexOf('-', dash2 + 1)
+    val dash4 = name.indexOf('-', dash3 + 1)
+    val dash5 = name.indexOf('-', dash4 + 1)
+
+    // For any valid input, dash1 through dash4 will be positive and dash5 will be negative,
+    // but it's enough to check dash4 and dash5:
+    // - if dash1 is -1, dash4 will be -1
+    // - if dash1 is positive but dash2 is -1, dash4 will be -1
+    // - if dash1 and dash2 is positive, dash3 will be -1, dash4 will be positive, but so will dash5
+    if (dash4 < 0 || dash5 >= 0) throw new IllegalArgumentException("Invalid UUID string: " + name)
+
+    var mostSigBits = parseLong(name, 0, dash1, 16) & 0xffffffffL
+    mostSigBits <<= 16
+    mostSigBits |= parseLong(name, dash1 + 1, dash2, 16) & 0xffffL
+    mostSigBits <<= 16
+    mostSigBits |= parseLong(name, dash2 + 1, dash3, 16) & 0xffffL
+
+    var leastSigBits = parseLong(name, dash3 + 1, dash4, 16) & 0xffffL
+    leastSigBits <<= 48
+    leastSigBits |= parseLong(name, dash4 + 1, len, 16) & 0xffffffffffffL
+
+    new java.util.UUID(mostSigBits, leastSigBits)
+  }
+
+  // Adapted from java.lang.Long.parseLong
+  @nowarn("msg=implicit numeric widening")
+  private def parseLong(s: CharSequence, beginIndex: Int, endIndex: Int, radix: Int): Long = {
+    if (s == null) throw new NullPointerException("input to parseLong is null")
+    if (beginIndex < 0 || beginIndex > endIndex || endIndex > s.length) throw new IndexOutOfBoundsException
+    if (radix < Character.MIN_RADIX)
+      throw new NumberFormatException("radix " + radix + " less than Character.MIN_RADIX")
+    if (radix > Character.MAX_RADIX)
+      throw new NumberFormatException("radix " + radix + " greater than Character.MAX_RADIX")
+    var negative = false
+    var i        = beginIndex
+    var limit    = -Long.MaxValue
+    if (i < endIndex) {
+      val firstChar = s.charAt(i)
```

Comment:
I ended up assigning everything other than hex characters in the `HextoDecimal` array to -1 so that it would trigger the following check:
```scala
if (digit < 0 || result < multmin) throw charSequenceError(s, beginIndex, endIndex, i)
```
Please let me know if you think it sufficient

Reply:
This was done in the OpenJDK implementation already - I missed it the first time around